### PR TITLE
submit_selected(): Don't override user's Referer

### PR DESCRIPTION
**Pull request by johnhawkinson**
(This is not intended to be squashed; I can break it into two PRs if that's better)

* prereq: Sort and group all imports per PEP8 (isort)
* actual: submit_selected(): Referer fixes

Don't override a Referer field that a user passes in.
Note that since HTTP headers are case-insensitive, use requests'
CaseInsensitiveDict to handle this.

Futhermore, don't modify the caller's **kwargs, the caller should
assume we do not modify the submitted dict, so make a copy
first.

Do this in _merge_referer() helper function, as we anticipate needing
this in other functions when #362 is fixed.

Add a test for overriding Referers.

This was raised in https://github.com/MechanicalSoup/MechanicalSoup/issues/362#issuecomment-812921602

### CODE REVIEW

**Review by hemberger — COMMENTED** [review]
This looks great, thank you for your contribution! I have a few minor questions/suggestions, but I think those should be resolved quite easily. I would like to give @moy an opportunity to weigh in as well, so I hope you don't mind if we have to wait a few days before merging this in.

File: mechanicalsoup/stateful_browser.py
Comment (johnhawkinson) [reply mechanicalsoup/stateful_browser.py]: Thanks for taking the time to demonstrate it very clearly, you're quite right.
I did a test of my own beforehand, but was improperly specified. Specifically, `kwargs` is a shallow copy so a function setting `kwargs['headers']['Referer']` would end up modifying the passed-in `headers` dict, but setting `kwargs['foobar']` does not change the passed-in `kwargs`.

So no need for the copy of `kwargs`, and we implicitly copy `headers` when we make the `CaseInsensitiveDict`.

I've marked this PR draft (never tried that before, what could go wrong?) and pushed 55b19b73702a3e2d358007393507ab00e354e195 to fix this. I'll squash it into c52e3d9 before finalizing the PR.

File: tests/test_stateful_browser.py
Comment (johnhawkinson) [reply tests/test_stateful_browser.py:449]: For (1), I actually thought about this (and did have such a test in my work tree), but it seemed excessive and I didn't feel like bothering to learn the ins and outs of `@pytest.mark.parametrize`, although it does look cool.  I'll add such.

For (2), well, given the confusion above, I guess it's not quite that — or perhaps it's a question of semantics. It'd be a test that a _reference_  referred to by `kwargs` is not modified. I guess in practice that could be any of `params`, `data`, `cookies`, `files`, `proxies`, or, of course, `headers`. I guess the test could `copy.deepcopy(kwargs)` beforehand and then compare each of those dicts afterwards with `==`. Seems a bit much, and making use of special knowledge of the guts of a requests `request`.  If https://stackoverflow.com/questions/25044073/comparing-python-objects-from-deepcopy is to be believed, doing this properly is, well, a bit much? Maybe? Do you think it's worth it?

**Review by moy — APPROVED** [review]
No strong opinion on ongoing discussions, but with or without the suggested changes, this looks good to me. Thanks for your contribution.

**Comment by johnhawkinson** [comment]
Alright, I believe these are addressed and I'll remove the draft marking.

Since 0eb9495 (isort) changes many files and is a prereq but not functionally related, it should be a separate commit, so this should be merge-commit-merged or rebase-merged rather than squash-merged.

If it aids in the review, the -wip version of this PR with the commits broken out from their state 2 days ago is available at https://github.com/MechanicalSoup/MechanicalSoup/compare/master...johnhawkinson:referer-override2-wip?expand=1

File: tests/test_stateful_browser.py
Comment (johnhawkinson) [reply tests/test_stateful_browser.py]: Done. As for (2), it's more straightforward than I thought, apparently `dict.__eq__()` recursively compares dicts, so it's easy enough to `copy.deepcopy()` and then compare afterwards.

I'll leave it to you to mark this and the other active section of the PR as Resolved, unless that's the wrong convention for this project.

**Review by hemberger — APPROVED** [review]

**Comment by hemberger** [comment]
Thanks again for the PR @johnhawkinson!

**Pull request merged**
